### PR TITLE
[Cache] Fix expiries handling

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php [hunk 1]
```diff
@@ -168,9 +171,12 @@ public function hasItem($key)
         if (isset($this->deferred[$key])) {
             $item = (array) $this->deferred[$key];
             try {
-                $e = null;
+                $e = $ok = null;
                 $value = $item[CacheItem::CAST_PREFIX.'value'];
-                $ok = $this->doSave(array($key => $value), $item[CacheItem::CAST_PREFIX.'lifetime']);
+                $lifetime = $item[CacheItem::CAST_PREFIX.'expiry'];
+                if (0 <= $lifetime = null !== $lifetime ? $lifetime - time() ?: -1 : 0) {
```
Comment (stloyd) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: This is quite unreadable IMO.

Comment (javiereguiluz) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: I don't know if others agree, but I find this line very hard to read+understand:

```php
if (0 <= $lifetime = null !== $lifetime ? $lifetime - time() ?: -1 : 0)
```

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: I wrote it an other way. It's really isomorphic so only a matter of personal taste. Please suggest the better way of writting this if the new proposed doesn't fits yours...

Comment (javiereguiluz) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: OK. Thanks. Just asking: could we get rid of the `null` value for `expiry` and use only numeric values? (using `-1` when necessary)

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: What would be the benefit? We still would have to deal with this special case (aka infinite expiry). Semantically, null is more appropriate to me, don't you think?

Comment (javiereguiluz) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: OK. Let's keep it like it's now. Thanks.

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php [hunk 2]
```diff
@@ -163,24 +166,10 @@ public function getItems(array $keys = array())
      */
     public function hasItem($key)
     {
-        $id = $this->getId($key);
-
-        if (isset($this->deferred[$key])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: this optim isn't worth it I think

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php [hunk 3]
```diff
@@ -286,44 +275,50 @@ public function saveDeferred(CacheItemInterface $item)
...
-                $ok = false;
             }
-            if (false === $ok) {
-                $ok = array_keys($values);
+            if (true === $e || array() === $e) {
+                continue;
             }
-            foreach ($ok as $id) {
-                $ko[$lifetime][] = array($id => $values[$id]);
+            if (is_array($e) || 1 === count($values)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: don't retry when only one item or when the adapter returned an explicit failing subset

File: src/Symfony/Component/Cache/Adapter/ArrayAdapter.php [hunk 4]
```diff
@@ -78,7 +78,7 @@ public function getItems(array $keys = array())
             $this->validateKey($key);
         }
 
-        return $this->generateItems($keys);
+        return $this->generateItems($keys, time());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/ArrayAdapter.php]: compare expiration at `getItems` time, returning a "frozen" result, as already discussed with @stof

**Comment by nicolas-grekas** [comment]
Tests are green
ping @symfony/deciders

**Pull request merged**